### ISSUE TITLE
Rule to detect duplicate function definitions.

### ISSUE THREAD

**Issue by ollz272** [astral-sh/ruff#10522]
We have the rule PIE794 which detects duplicate class field definitions. I think it would be nice to have a rule which detects duplicate function definitions:

```python
# Error
class Foo:
    
    def bar():
        print("hello world")

    def bar():
        print("hello ruff")
```

This is clearly a mistake and would be great if it was caught by ruff.

**Comment by MichaReiser** [comment]
Have you tried `F811`? It flags the second declaration of `bar` https://play.ruff.rs/f029b8b6-26bd-49f9-9da2-ac2b22685b0b

**Comment by kaddkaka** [comment]
For function in file scope the duplicated function is found by both pylint `E0102` and mypy `no-redef`, example:
```py def banana():
    """banana"""


def banana():
    """banana"""
```

And in the class method duplication case:
```py
lass BananaTree:
   """BananTree"""

   def banana(self):
       """banana"""

   def banana(self):
       """banana"""
```

Shows these errors:
```
pylint: [function-redefined] method already defined line 4 [E0102]
pyflakes: redefinition of unused 'banana' from line 4
```

According to https://github.com/astral-sh/ruff/issues/970 Pylint rule `E0102` is covered by `F811` that @MichaReiser mentioned. 

That info is missing but should be added to documentation at https://docs.astral.sh/ruff/rules/#pylint-pl

**Comment by kaddkaka** [comment]
`F811` is however called [redefined-while-unused](https://docs.astral.sh/ruff/rules/redefined-while-unused/) and only triggers when the first definition has not been called. This means that F811 is easily missed for exampled in this example:

```py
def banana():
    """banana"""


def banana():
    """banana"""
```

Give no errors from pyflakes at least. Since the first definition is used, it doesn't report the redefinition as an error. I'm not sure what ruff does in this case.

**Comment by ollz272** [comment]
I see why this isn't being picked up for us, we have private variables, which it doesn't work for
https://play.ruff.rs/ee3c213f-21d7-4540-a052-9acbc6ac4d6f

**Comment by charliermarsh** [comment]
Why should the second definition be flagged here?

```python
def banana():
    """banana"""

banan()

def banana():
    """banana"""
```

**Comment by charliermarsh** [comment]
Yeah, that's due to [`dummy-variable-rgx`](https://docs.astral.sh/ruff/settings/#lint_dummy-variable-rgx). I wonder if that shouldn't be applied to functions. I'm not really sure. It's possible that users would find that really surprising.

**Comment by kaddkaka** [comment]
> Why should the second definition be flagged here?
> 
> ```python
> def banana():
>     """banana"""
> 
> banana()
> 
> def banana():
>     """banana"""
> ```
> 
> This seems valid to me.

I didn't state any opinion about the rule, but E0102 *does* trigger on this code.

Opinion: Sure this could be a valid case, but it's probably quite rare to redefine functions like this. I can see programmers prefer to have a warning for this. If one doesn't want the warning it can be turned off.
